pyhacrf/algorithms.pyx

```diff
@@ -30,9 +30,10 @@ cpdef dict forward(list lattice, np.ndarray[double, ndim=3] x_dot_parameters):
     return alpha
 
 
-cpdef dict forward_predict(list lattice, np.ndarray[double, ndim=3] x_dot_parameters):
+cpdef np.ndarray forward_predict(list lattice, np.ndarray[double, ndim=3] x_dot_parameters):
     """ Helper to calculate the forward weights.  """
-    cdef dict alpha = {}
+    cdef np.ndarray[double, ndim=3] alpha = np.empty_like(x_dot_parameters)
```

Comment:
Ah I see this is the trick - the fact that the `forward_predict` function doesn't need to store the edge potentials makes it possible to store the `alpha`s in an array of the same shape as `x_dot_parameters`. Cool!